### PR TITLE
DOCSP-32693: FindOne by ObjectId

### PR DESCRIPTION
**Pull request by norareidy**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-32693>
Staging - <https://docs-mongodbcom-staging.corp.mongodb.com/golang/docsworker-xlarge/DOCSP-32693-find-by-id/fundamentals/crud/read-operations/retrieve/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 1]
```diff
@@ -203,6 +213,56 @@ to the ``FindOne()`` method, which performs the following actions:
...
+      :visible: false
+
+      [{"Key":"item","Value":"Hibiscus"},{"Key":"rating","Value":4}]
+
+.. note::
+
+   The {+driver-short+} automatically generates a unique ``ObjectId`` value for
+   each document's ``_id`` field. The preceding code example uses an ``ObjectID``
+   with a hexidecimal string representation of ``"65170b42b99efdd0b07d42de"``,
+   but your ``ObjectID`` values are unique.
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:264]: S: this note would be more helpful if it linked to some information about the `_id` field, like [this](https://docs-mongodbcom-staging.corp.mongodb.com/golang/docsworker-xlarge/DOCSP-32693-find-by-id/fundamentals/crud/write-operations/insert/#the-_id-field). I think the message of this sentence is just that users will have to know the Id of the doc they are looking for to search with this function -- and that message is kind of confusing right now

**Review by rustagir — CHANGES_REQUESTED** [review]

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 2]
```diff
@@ -203,6 +213,56 @@ to the ``FindOne()`` method, which performs the following actions:
...
+      )
+
+      var result bson.D
+      err = coll.FindOne(context.TODO(), filter).Decode(&result)
+      if err != nil {
+          panic(err)
+      }
+
+      res, _ := json.Marshal(result)
+      fmt.Println(string(res))
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/fundamentals/crud/read-operations/retrieve.txt]: Moved!

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 3]
```diff
@@ -75,7 +77,13 @@ The ``FindOne()`` method expects you to pass a ``Context`` type and a
 query filter. The method returns *the first document* that matches the
 filter as a ``SingleResult`` type.
 
-To learn how to access data in a ``SingleResult`` see :ref:`golang-bson-unmarshalling`.
+For a ``FindOne()`` method code example, see the :ref:`golang-find-one-example`
+section of this page. For a ``FindOne()`` code example that queries by using
+a specific ``ObjectId`` value, see the :ref:`golang-find-one-by-id` section
+of this page.
+
+To learn how to access data in a ``SingleResult``, see
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:85]:
```suggestion
To learn how to access data from a ``SingleResult`` type, see
```

Comment (rustagir) [reply source/fundamentals/crud/read-operations/retrieve.txt:85]: In general, where a type/method is mentioned, explain what it is. Aka 
the ``UpdateOne()`` method
a value of type ``Result``
an instance of ``MyStruct``

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 4]
```diff
@@ -180,28 +171,70 @@ to the ``FindOne()`` method, which performs the following actions:
...
 
-      filter := bson.D{{"date_ordered", bson.D{{"$lte", time.Date(2009, 11, 30, 0, 0, 0, 0, time.Local)}}}}
-      opts := options.FindOne().SetSkip(2)
-      
-      var result Review
-      err := coll.FindOne(context.TODO(), filter, opts).Decode(&result)
-      if err != nil {
-          panic(err)
+This example defines an ``id`` variable with a value of type ``ObjectId``
+and specifies a query filter with the ``id`` variable. The value of ``id``
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:194]: S: avoid with where possible - [sg](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#std-term-using:~:text=word%2C%20username.-,using,-by%20using%20with)
```suggestion
This example defines an ``id`` variable with a value of type ``ObjectId``
and specifies a query filter that matches documents with an ``_id`` field that matches this variable.
The value of ``id``
```

Comment (rustagir) [reply source/fundamentals/crud/read-operations/retrieve.txt:194]: Then, I feel like the first sentence is sufficient to delete the second sentence.

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 5]
```diff
@@ -180,28 +171,70 @@ to the ``FindOne()`` method, which performs the following actions:
...
+      {
+         _id: ObjectId('65170b42b99efdd0b07d42de'),
+         item: "Hibiscus",
+         rating : 4,
+         date_ordered : 2009-12-18T05:00:00.000+00:00
       }
-      
-      res, _ := json.Marshal(result)
-      fmt.Println(string(res))
+   ]
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:208]: S: remove these outer square brackets and dedent

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 6]
```diff
@@ -180,28 +171,70 @@ to the ``FindOne()`` method, which performs the following actions:
...
+
+The ``FindOne()`` method recieves the filter and a ``FindOneOptions``
+instance as parameters and performs the following actions:
+
+- Matches the document with the specified ``ObjectId`` value
+- Projects the ``Item`` and ``Rating`` fields of the matched document
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:214]:
```suggestion
- Match the document with the specified ``ObjectId`` value
- Project only the ``Item`` and ``Rating`` fields of the matched document
```

**Review by rustagir — CHANGES_REQUESTED** [review]
a few more things - lmk if you have questions about the struct comment

File: source/fundamentals/crud/read-operations/retrieve.txt (hunk 4 above)
Comment (norareidy) [reply source/fundamentals/crud/read-operations/retrieve.txt:194]: Slightly different wording, but fixed

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 7]
```diff
@@ -180,29 +171,70 @@ to the ``FindOne()`` method, which performs the following actions:
...
+      :start-after: begin objectid
+      :end-before: end objectid
+      :language: go
+      :dedent:
+
+  .. output::
+      :language: none
+      :visible: false
+
+      {"item":"Hibiscus","rating":4}
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:228]: Q: this output shows the field names as lowercase, but if the code is unmarshalling the document to the `Review` type, it should look different, as it should print with the fieldnames from the struct.

File: source/fundamentals/crud/read-operations/retrieve.txt [hunk 8]
```diff
@@ -180,29 +171,70 @@ to the ``FindOne()`` method, which performs the following actions:
...
+
+.. note::
+
+   The {+driver-short+} automatically generates a unique ``ObjectId``
+   value for each document's ``_id`` field, so your ``ObjectId`` value
+   will differ from the preceding code example. For more information
```
Comment (rustagir) [inline source/fundamentals/crud/read-operations/retrieve.txt:234]:
```suggestion
   might differ from the preceding code example. For more information
```

**Review by rustagir — APPROVED** [review]
lgtm with two small items! Great job, I know the code with going back and forth b/t bson and go is hard :)

File: source/fundamentals/crud/read-operations/retrieve.txt (hunk 7 above)
Comment (norareidy) [reply source/fundamentals/crud/read-operations/retrieve.txt:228]: When I run the code, it outputs the lowercase field names so maybe it's not unmarshalling to a `Review`? I might be missing something in the code

Comment (norareidy) [reply source/fundamentals/crud/read-operations/retrieve.txt:228]: I think the field names will stay lowercase after using the MarshalExtJSON() function, unless I unmarshal again into the struct. Looks like the printed documents on[ this page](https://www.mongodb.com/docs/drivers/go/current/fundamentals/crud/read-operations/project/) also have lowercase fields.

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Pull request merged**